### PR TITLE
add mkldnn surport for concat

### PR DESCRIPTION
**Pull request by wentingj**
## Description ##
Add mkldnn surport for concat, which passed unit test. First epoch of inception-bn has "Epoch[0] Train-accuracy=0.573438"

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/concat.cc [hunk 1]
```diff
@@ -103,12 +104,104 @@ static bool ConcatType(const nnvm::NodeAttrs& attrs,
...
+}
+
 struct ConcatGrad {
   const char *op_name;
   std::vector<nnvm::NodeEntry> operator()(const nnvm::NodePtr& n,
                                           const std::vector<nnvm::NodeEntry>& ograds) const {
-    const ConcatParam& param = nnvm::get<ConcatParam>(n->attrs.parsed);
+    CHECK_EQ(ograds.size(), 1);
     std::vector<nnvm::NodeEntry> heads(ograds.begin(), ograds.end());
+    for (size_t i = 0; i < n->inputs.size(); i++) {
```
Comment (zheng-da) [inline src/operator/nn/concat.cc:198]: this makes native concat fail. The native concat only expects out_grad.

**Comment by zheng-da** [comment]
@wentingj could you run "make lint" and fix the coding style, please? Thanks.
I just pushed new commits to fix the coding style in my code.

**Comment by wentingj** [comment]
@zheng-da Thanks for your kind reminder. I have run "make lint" and passed coding style checking in my branch "wentingj:mkldnn-concat".

**Review by zheng-da — APPROVED** [review]
